### DISCUSSION TITLE
Monorepo project

### DISCUSSION THREAD

**Discussion by DhivinX** [xeokit/xeokit-sdk#1414] [Ideas]
Hello, 

I have been observing and analyzing this repository since the beginning of the week and wondering what would be the best way to convert it into a monorepo.  I need help understanding:

Are any external repositories required for this to work?

I wonder if it is possible to convert them into a monorepo in a reasonable time, many changes could be breaking, are there specific reasons for this?

I'm trying to understand what could be included in `packages`, I would need a description or diagram of what would be included there.

I can't think of anything else at the moment. First of all, I need alignment from you @xeolabs

**Comment by paireks** [comment]
Just a note here to remember:

switching here a default branch from main branch to develop (same as what was done in v3) is a little bit breaking. In v3 it doesn't matter that much cause it's alpha-stage, but here there are 262 forks already. Most of the forks are "main" clone only probably (cause it is a default way to fork). So everyone will need to update their fork to take "develop" branch as well. I'm telling this, because I had this issue with v3 change, I've seen updates in the main repository, used Github functionality to update my fork, but couldn't find the reason why I can't update them, to find out some time later the default branch has switched :) Such change here would require some form of announcement at least.

**Reply by DhivinX** [reply]
I agree with you 100%, but I have an idea to rename the `main` branch to `develop` and then add the `main` branch again. This is a workaround in my opinion, but of course treat it with caution.

**Comment by paireks** [comment]
There was a time I was drawing some form of diagrames to find out where is what and why, but it's WIP, and I'm not sure if it could be helpful there. But will be happy to share if needed

![2024-03-14_14h53_43](https://github.com/xeokit/xeokit-sdk/assets/47977819/808a3e89-a5ef-40ec-9303-6111cf61d6f3)

**Reply by DhivinX** [reply]
Thanks @paireks, good for a start! Could you please send a screenshot in a higher resolution? If it's possible.

**Reply by paireks** [reply]
@DhivinX Here it is, just load it in https://app.diagrams.net/. I did it for myself initially, but maybe with your input we could make it a part of documentation in the future, so please edit it according to your needs :)
[Xeokit.zip](https://github.com/xeokit/xeokit-sdk/files/14607953/Xeokit.zip)

**Comment by xeolabs** [comment]
> Are any external repositories required for this to work?

See below..
 
> I wonder if it is possible to convert them into a monorepo in a reasonable time, many changes could be breaking, are there specific reasons for this?

Main reason would be that currently we bundle everything, viewer and plugins, into the same binaries, which includes a bunch of 3rd-party libraries that support various plugins. For example, we have various loader plugins, like `GLTFLoaderPlugin`, which use third-party libraries, like `loaders.gl`, to parse their various file formats. 

This means that these 3rd-party libraries are bundled with the binary regardless of whether users use them, and that user's apps have to have those 3rd-party libs as npm dependencies, even if not using them.

But then, some users just import the source files straight out of `@xeokit-sdk/src`, so thinking this through, actually I'm not 100% sure that making this a monorepo is even worth it the hassle! 

The 3rd-party dependencies are only these: 

`
  "dependencies": {
    "@loaders.gl/core": "^3.2.6",
    "@loaders.gl/gltf": "^3.2.6",
    "@loaders.gl/las": "^3.2.6",
    "html2canvas": "^1.4.1"
  },

which is not a lot..

> I'm trying to understand what could be included in `packages`, I would need a description or diagram of what would be included there.

Just briefly for this discussion: I was thinking that there would be a package for each plugin in `./src/plugins`, eg. `@xeokit/xeokit-gltf-loader`, and a package for the Viewer in `./src/viewer`, eg `@xeokit/xeokit-viewer`. Perhaps also a package that deploys to @xeokit/xeokit-sdk` and pulls together the source from all the other packages, which could ensure backward compatibility for users.

> I can't think of anything else at the moment. First of all, I need alignment from you @xeolabs

Now I'm starting to wonder if it's overkill to make this a monorepo, since `@xeokit/sdk` is the next version and that's already a monorepo, with good modularity..

**Reply by DhivinX** [reply]
Yes it's true, xeokit/sdk is a perfect example for using monorepo. I have the same doubts as you, and sooner or later you will switch to v3, while v2 will become the LTS version. What I'm most afraid of when modifying version v2 is the possible lack of backward compatibility with already working applications, as these changes may be breaking. In any case, the decision is yours, think it over carefully, in the meantime I can start working on examples of using xeokit-bim-viewer, see https://github.com/xeokit/xeokit-sdk/discussions/1417
